internal/perms/grants.go

```diff
@@ -55,7 +55,11 @@ type GrantTuples []GrantTuple
 // GrantsHash returns a stable hash of all the grants in the GrantTuples.
 func (g GrantTuples) GrantHash(ctx context.Context) ([]byte, error) {
 	const op = "perms.(GrantTuples).GrantHash"
-	// TODO: Should this return an error when the GrantTuples is empty?
+
+	if len(g) == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "empty GrantTuples receiver passed into GrantHash")
+	}
```

Comment:
So, currently, we return a slice of empty bytes when there are no grant tuples? I'm not sure if that is wrong, at least it's a unique mapping from the input to the output. As long as the slice isn't actually empty, that should keep working, right? What happens if we stop making this an error?

Reply:
Yes

Technically there's a non-nil input of GrantTuples that'll hash to a slice of empty bytes (possible but very unlikely). So by also mapping nil GrantTuples to the empty bytes slice, this function would _technically_ no longer produce unique hashes

If we stop making this an error & return nil, there will be an error in some of the pagination functions (`List`, `ListPage`, etc) where we check for `len(grantsHash) == 0`. If we return the empty byte slice instead, things will work normally, but I guess there would be the possibility of some non-nil GrantTuple input hashing to an empty byte slice

Reply:
I personally think we're OK with the remote possibility of a valid grants combination hashing to 32 bytes of 0s as a collision in this case. I'm tempted to say we should keep mapping (no grants) -> (0000000). Does that sound OK?

Reply:
@johanbrandhorst That sounds good to me 👍

Reply:
My opinion on this is that a `nil` or empty `GrantsTuples` should still be usable. So calls to `Hash` should not result in an error in either case.

Reply:
Reverted back to normal & added a comment clarifying this behavior: [00550a0](https://github.com/hashicorp/boundary/pull/5334/commits/00550a064a070ee1db89ce9b0e8728e0205d1d1d)